python/device/device.py

```diff
@@ -514,34 +482,94 @@ def pairing_thread(self, worker_thread_info):
                 elif (
                     time.time() - pairing_last_request_epochtime
                 ) > self.config.pairing_request_send_interval_in_seconds:
+                    logger.info("pairing response timeout.  Requesting again")
+                    repeat_pairing_request = True
 
-                    logger.info("sending pairingRequest message")
-                    pairing_request = {
-                        "thief": {
-                            "cmd": "pairingRequest",
+            # We trigger a pairing operation by pushing this string into our incoming queue.
+            # Not a great design, but it lets us group this functionality into a single thread
+            # without adding another signalling mechanism.
+            if msg == "START_PAIRING":
+                pairing_start_epochtime = time.time()
+            if msg == "START_PAIRING" or repeat_pairing_request:
+                # Set our reported properties to start a pairing operation or to try again if no
+                # service has responded yet.
+                logger.info("Starting pairing operation")
+                currently_pairing = True
+                self.pairing_complete = False
+                pairing_last_request_epochtime = time.time()
+                self.pairing_id = str(uuid.uuid4())
+                self.service_run_id = None
+                props = {
+                    "thief": {
+                        "pairing": {
                             "requestedServicePool": requested_service_pool,
+                            "serviceRunId": None,
+                            "pairingId": self.pairing_id,
+                            "deviceRunId": run_id,
                         }
                     }
-                    msg = self.create_message_from_dict(pairing_request)
-                    self.outgoing_control_message_queue.put(msg)
-
-                    pairing_last_request_epochtime = time.time()
+                }
+                logger.info("updating pairing reported props: {}".format(pprint.pformat(props)))
+                self.client.patch_twin_reported_properties(props)
+                azure_monitor.add_logging_properties(pairing_id=self.pairing_id)
 
-            elif msg:
-                if currently_pairing:
-                    thief = json.loads(msg.data.decode()).get("thief")
-                    if thief["cmd"] == "pairingResponse":
-                        service_app_run_id = thief["serviceAppRunId"]
-                        logger.info(
-                            "Service app {} claimed this device instance".format(service_app_run_id)
+            elif msg and isinstance(msg, dict):
+                # is msg is a dict, that means we have a desired property change.  A service
+                # might be trying to pair with us.
+                logger.info("received pairing desired props: {}".format(pprint.pformat(msg)))
```

Comment:
Why not base the pairing request off a message property?

Reply:
I'm documenting this now.  The pairing process is a handshake with 4 steps, all done with reported and desired properties:                                

1. A device app says "I need a partner" by setting it's reported `serviceRunId` to `null`.                                     
2. A service app says "I'm available" by setting the devices desired `serviceRunId` to the service app's `runId` value.        
3. The device app says "I choose you" by setting it's reported `serviceRunId` to the `runId` of the chosen service             
4. The service app finalizes the pairing by setting the devices desired `acceptedPairing` value.                               
                                                                                                                               
Once `acceptedPairing` is set, the pairing is complete and test operation can begin.

Reply:
1-3 make sense to me. Why is 4 necessary?

Reply:
Because of network latency and out-of-order delivery.  
The device app needs to know when it can start "regular operation" 
After step 3, the device app knows that it sent the "pairing is complete" message, but it doesn't know that the service app received it, and it doesn't know if the service app is "ready for regular traffic".

Reply:
Hmm, okay. I'll buy that.

Back to my original question, why do we figure out the type of message based on the type (dictionary)? That seems like a weak connection.

Reply:
Ah, I misunderstood your question.  You're right.  This was not done well.  I'm going to revisit this in a later commit.